core/schema/moddag.go

```diff
@@ -0,0 +1,523 @@
+package schema
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/dagger/dagger/core"
+	"github.com/moby/buildkit/util/bklog"
+	"github.com/opencontainers/go-digest"
+	"github.com/vektah/gqlparser/v2/formatter"
+)
+
+const (
+	modMetaDirPath    = "/.daggermod"
+	modMetaOutputPath = "output.json"
+
+	coreModuleName = "daggercore"
+)
+
+/*
+Mod is a module in loaded into the server's DAG of modules; it's the vertex type of the DAG.
+It's an interface so we can abstract over user modules and core and treat them the same.
+*/
+type Mod interface {
+	// The name of the module
+	Name() string
+
+	// The digest of the module itself plus the recursive digests of the DAG it depends on
+	DagDigest() digest.Digest
+
+	// The direct dependencies of this module
+	Dependencies() []Mod
+
+	// The schema+resolvers exposed by this module (does not include dependencies)
+	Schema(context.Context) ([]SchemaResolvers, error)
+
+	// The introspection json for this module's schema
+	SchemaIntrospectionJSON(context.Context) (string, error)
+
+	// ModTypeFor returns the ModType for the given typedef based on this module's schema. If checkDirectDeps is true,
+	// then its direct dependencies will also be checked.
+	ModTypeFor(ctx context.Context, typeDef *core.TypeDef, checkDirectDeps bool) (ModType, bool, error)
```

Comment:
I tend to prefer the extra `bool` in cases where it's expected that a value may not be found and you want all callers to handle it, since nil checks being required can quickly turn into nil panics in prod.

Reply:
When `checkDirectDeps` is true, how does that relate to namespacing? I'm guessing the queried `typeDef` is already namespaced to `FooType` for a module `Foo`?